lib/frame.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class Frame
+  TEN_PINS = 10
+  FIRST_ROLL = 0
+  SECOND_ROLL = 1
+
+  def initialize
+    @rolls = []
+    @frame_type = nil
+  end
+
+  def log_roll(pins_downed)
+    raise 'Pins downed must be between 0 and 10' if invalid_pins?(pins_downed)
+    raise 'Frame complete. Cannot roll again' if frame_complete?
+
+    process_frame_type(pins_downed)
+    process_roll(pins_downed)
+  end
+
+  def sum_frame
+    @rolls.sum
+  end
+
+  def all_rolls
+    @rolls
+  end
+
+  def frame_complete?
+    @rolls.count > 1 || @frame_type == :strike
+  end
+
+  def strike_frame?
+    @frame_type == :strike
+  end
+
+  def spare_frame?
+    @frame_type == :spare
+  end
+
+  def open_frame?
+    @frame_type == :open
+  end
+
+  private
+
+  def invalid_roll?(roll)
+    roll.negative? || (roll > @rolls.count)
+  end
+
+  def invalid_pins?(pins_downed)
+    pins_downed.negative? || pins_downed > TEN_PINS
+  end
+
+  def to_many_pins?(pins_downed)
+    return false if @frame_type == :strike
+
+    pins_downed > pins_left
+  end
+
+  def pins_left
+    TEN_PINS - @rolls[FIRST_ROLL]
+  end
+
+  def process_roll(pins_downed)
+    process_second_roll(pins_downed)
+    process_first_roll(pins_downed)
+  end
+
+  def process_second_roll(pins_downed)
+    return unless second_roll?
```

Comment:
It would be better to move this conditional into the `process_roll` method, so that you're only ever calling `process_second_roll` when you are processing a second roll.

Reply:
implemented in the javascript version.  Although not finished I don't think I need to concern myself about 1st and second rolls now as I can now just look at the isStrike/Spare methods